### PR TITLE
[jit] Transition some memory allocations to memory managers.

### PR DESCRIPTION
**Pull request by vargaz**

### CODE REVIEW

**Comment by srxqds** [comment]
why no replace in aot–complier or aot-runtime?

File: mono/metadata/class-internals.h [hunk 1]
```diff
@@ -1592,6 +1592,59 @@ m_field_get_offset (MonoClassField *field)
...
 }
 
+/*
+ * Memory allocation for classes/methods
+ *
+ *   These should be used to allocate memory whose lifetime is equal to
+ * the lifetime of the domain+class/method pair.
+ */
+
+static inline MonoMemoryManager*
```
Comment (CoffeeFlux) [inline mono/metadata/class-internals.h:1602]:
```suggestion
static inline MonoMemoryManager *
```

File: mono/metadata/class-internals.h [hunk 2]
```diff
@@ -1592,6 +1592,59 @@ m_field_get_offset (MonoClassField *field)
...
+	return mono_mem_manager_alloc (m_class_get_mem_manager (domain, klass), size);
+}
+
+static inline void *
+m_class_alloc0 (MonoDomain *domain, MonoClass *klass, guint size)
+{
+	return mono_mem_manager_alloc0 (m_class_get_mem_manager (domain, klass), size);
+}
+
+static inline MonoMemoryManager*
```
Comment (CoffeeFlux) [inline mono/metadata/class-internals.h:1625]:
```suggestion
static inline MonoMemoryManager *
```

File: mono/metadata/class-setup-vtable.c [hunk 3]
```diff
@@ -1967,7 +1967,9 @@ mono_class_setup_vtable_general (MonoClass *klass, MonoMethod **overrides, int o
 		MonoMethod *impl = overrides [i*2 + 1];
 		if (!MONO_CLASS_IS_INTERFACE_INTERNAL (decl->klass)) {
 			g_assert (decl->slot != -1);
+#ifdef FEATURE_COVARIANT_RETURNS
```
Comment (CoffeeFlux) [inline mono/metadata/class-setup-vtable.c:1970]: This seems unrelated, I'd rather move it out to another PR (and ping @lambdageek).

File: mono/metadata/loader-internals.h [hunk 4]
```diff
@@ -241,4 +241,7 @@ mono_mem_manager_code_commit (MonoMemoryManager *memory_manager, void *data, int
 void
 mono_mem_manager_code_foreach (MonoMemoryManager *memory_manager, MonoCodeManagerFunc func, void *user_data);
 
+char*
```
Comment (CoffeeFlux) [inline mono/metadata/loader-internals.h:244]:
```suggestion
char *
```

File: mono/metadata/memory-manager.c [hunk 5]
```diff
@@ -169,6 +169,17 @@ mono_mem_manager_alloc0_nolock (MonoMemoryManager *memory_manager, guint size)
 	return mono_mempool_alloc0 (memory_manager->mp, size);
 }
 
+char*
```
Comment (CoffeeFlux) [inline mono/metadata/memory-manager.c:172]:
```suggestion
char *
```

File: mono/mini/Makefile.am.in [hunk 6]
```diff
@@ -770,7 +770,7 @@ libmonosgen_2_0_la_LIBADD = libmini.la $(interp_libs_with_mini) $(dbg_libs_with_
 libmonosgen_2_0_la_LDFLAGS = $(libmonoldflags) $(monobin_platform_ldflags) $(CCLDFLAGS)
 libmonosgen_2_0_la_LIBTOOLFLAGS = $(MONO_LIBTOOL_TAG)
 
-noinst_LIBRARIES += libmaintest.a
+#noinst_LIBRARIES += libmaintest.a
```
Comment (CoffeeFlux) [inline mono/mini/Makefile.am.in:773]: Again, is this related? I know Bernhard added the interpreter regression testing stuff, are we now getting rid of it?

File: mono/mini/interp/transform.h [hunk 7]
```diff
@@ -167,6 +167,7 @@ typedef struct
 	InterpBasicBlock *entry_bb, *cbb;
 	int bb_count;
 	MonoMemPool     *mempool;
+	MonoMemoryManager *mem_manager;
```
Comment (CoffeeFlux) [inline mono/mini/interp/transform.h:170]: Is it safe to add this reference? Are transforms tied to the lifetime of an assembly or is there any chance of one still existing after an ALC is unloaded? If you're sure it's safe I'm a big fan, this makes things much cleaner.

File: mono/mini/interp/transform.c [hunk 8]
```diff
@@ -1046,9 +1046,8 @@ static void
 interp_generate_ipe_throw_with_msg (TransformData *td, MonoError *error_msg)
 {
 	MonoJitICallInfo *info = &mono_get_jit_icall_info ()->mono_throw_invalid_program;
-	MonoMemoryManager *memory_manager = mono_domain_ambient_memory_manager (td->rtm->domain);
 
-	char *msg = mono_mempool_strdup (memory_manager->mp, mono_error_get_message (error_msg));
+	char *msg = mono_mem_manager_strdup (td->mem_manager, mono_error_get_message (error_msg));
```
Comment (CoffeeFlux) [inline mono/mini/interp/transform.c:1050]: This adds locking where it wasn't present previously, is that intended?

File: mono/mini/mini.h [hunk 9]
```diff
@@ -1349,6 +1349,8 @@ typedef struct {
 
 	MonoGSharedVtMethodInfo *gsharedvt_info;
 
+	MonoMemoryManager *mem_manager;
```
Comment (CoffeeFlux) [inline mono/mini/mini.h:1352]: Similar questions to adding it in the transform, but again if you're sure it's safe it makes for a very nice diff.

**Review by CoffeeFlux — APPROVED** [review]
Thanks so much!

File: mono/metadata/class-setup-vtable.c (hunk 3 above)
Comment (lambdageek) [reply mono/metadata/class-setup-vtable.c:1970]: It's just fixing an unused variable warning in mono/mono, I think.  I'm ok keeping it here

Comment (CoffeeFlux) [reply mono/metadata/class-setup-vtable.c:1970]: Oh it's not a functional change? Gotcha.

File: mono/metadata/class-internals.h (hunk 1 above)
Comment (vargaz) [reply mono/metadata/class-internals.h:1602]: We usually don't put spaces between those, only in variables, not in return values.

File: mono/mini/Makefile.am.in (hunk 6 above)
Comment (vargaz) [reply mono/mini/Makefile.am.in:773]: Unrelated, will fix.

File: mono/mini/interp/transform.h (hunk 7 above)
Comment (vargaz) [reply mono/mini/interp/transform.h:170]: TransformData is just a transient data structure used by transform.c, its the same as MonoCompile for the JIT.

File: mono/mini/interp/transform.c (hunk 8 above)
Comment (vargaz) [reply mono/mini/interp/transform.c:1050]: Having extra locking is only a problem during steady state execution, the transform code is part of startup.
